### DISCUSSION TITLE
Retrieving a refresh token

### DISCUSSION THREAD

**Discussion by tredpath** [zitadel/zitadel#7731] [Q&A] [answered]
I'm trying to get an access token and refresh token pair. The following settings will successfully return an access token but even with offline_access in the scope it doesn't return a refresh token.
![image](https://github.com/zitadel/zitadel/assets/1920516/464f40d6-0938-43c4-9a17-ec63d1399cd8)

If I turn on the checkbox for refresh token the call to /oauth/v2/token by my service times out. I saw errors `QUERY-WRFG4` and `OIDC-AhX2u` in the Zitadel logs from around the times of two different login attempts but I'm not sure if they're related or what they mean. What settings are other people using to get a refresh token, or is there something else wrong with my instance?

**Comment by dakshitha** [comment, marked as answer]
Technically, adding “offline_access” in the scope parameters to the "/oauth/v2/authorize" endpoint with the "Refresh Token" checkbox selected should work. I tried with ZITADEL Cloud and it works. Maybe it's something misconfigured in the request? I used Postman. 

<img width="800" alt="image" src="https://github.com/zitadel/zitadel/assets/11885043/f493588b-556a-4898-93ac-5ba0f625eb78">

**Reply by tredpath** [reply]
I might have just been overthinking this one. I increased the read timeout on the request and it does eventually return a good refresh token. The response time is considerably longer if a refresh token is requested than just an access token.

**Reply by fforootd** [reply]
Hm, do you self-host?

The main thing that influences latency is the clients authentication.

**Reply by tredpath** [reply]
Self-hosted, v2.44.0, running 3 nodes.

I did some quick timing of the /oauth/v2/token call in Postman with the Refresh Token setting on and off. The only difference between the calls was the code_verifier and code changed for each request.
`https://url.com/oauth/v2/token?code=retrieved-code&redirect_uri=https://url.com&grant_type=authorization_code&code_verifier=generated-verifier&client_id=placeholderclientid@test`

Without refresh tokens the response times averaged 0.676s, with refresh tokens the response times averaged 11.475s.

**Reply by fforootd** [reply]
Hm that is definitely a way to high latency. It should be below 2s.

We are currently working on perf improvements on the token endpoint, but they have not been released.

What you could check is in the CRDB transaction/query view if you see a high latency happen for some queries.
